### ISSUE TITLE
Stuck at "Updating search engine..."

### ISSUE THREAD

**Issue by timorl** [mailpile/Mailpile#2003]
After an update a couple days ago my mailpile started becoming unresponsive a short while after being started. By unresponsive I mean clicking on any email in any search result creates a (semi?)permanent "Working...." notification and does nothing else. Similarily using the 'view' command shows 'Parsing metadata for X results (full_threads=True)' and never returns (no prompt). In this state I wasn't able to turn off mailpile in any other way than sending it a TERM signal.

I'm not really getting any error messages anywhere, but there is a permanent "Updating search engine for..." notification in the web interface and it's also in the ongoing event log. The log contents are here: https://pastebin.com/chyMXHkA (some email addresses redacted).

I tried updating mailpile again just a couple hours ago and it seemed to help for a bit -- for about half an hour everything was working smoothly. Then the "Updating search...." message appeared and it's back to unusable.

Any tips on what I can try doing are appreciated.

**Comment by BjarniRunar** [comment]
Interesting, thank you for reporting this.

Things to try (at the `mailpile>` prompt):

1. `set sys.debug = log sources`
2. restart Mailpile: `quit restart`
3. `login`
4. `rescan sources`

... watch to see if any errors or stack traces or other strange things appear on the screen.

**Comment by timorl** [comment]
So, at point 2. I already got some stack traces: https://pastebin.com/BHGGkFbP . Generally stuff connected with writing to the history file, so possibly not connected? Also, if this was supposed to restart the app, not just quit it, then it didn't work.

Afterwards I started mailpile again and ran points 3. and 4. It didn't seem to do aything abnormal, but it did not finish after about 30minutes (in the sense that there was no prompt). In the meantime I also tried:

1. Running a `search` command (to check whether there is actually no prompt or if it is just not visible for some reason). No effect.
2. Logging in through the web interface, to check whether the symptoms are the same. It still has the "Updating search engine..." event going with the same status as in the previous paste. Now I noticed though that some of the emails can be read, only a some seem to be "broken", i.e. not displaying anything when clicking on them and creating a permanent "Waiting" message.

After that I had to go and turn off the machine. I sent an interrupt in the CLI, got a result and prompt so I typed `quit`. This got stuck at "Starting: Save config", so I killed the process (in the bug report I actually made a mistake, I don't send a TERM, but a KILL signal; the former doesn't do anything). The complete log is here: https://pastebin.com/7Ek7n63d .

**Comment by timorl** [comment]
I am terribly sorry to be impatient, but sending unsigned mail and using mail clients less nice than Mailpile bothers me. As this bug seems to at least stem from a data corruption issue I'd like to ask if there are some files I could easily delete that could be recreated/redownloaded? I strongly suspect that deleting the whole `.local/share/Mailpile` directory (or even just the `mail` subdirectory) would remedy the issue, but redownloading all my email would take quite a while. Is there a way to only delete a part of all the emails, starting with the newest ones?

I can also make a backup of the current state of the files, so one can debug it later.

**Comment by BjarniRunar** [comment]
Sorry, I have been travelling and working on other things. Unfortunately, this requires some thought for me to give useful advice, and I just haven't had time to focus on this.

Yes, deleting your `.local/share/Mailpile` will let you start from scratch (or just renaming `..../Mailpile/default` to `..../Mailpile/broken`). Just deleting the `mail` subdirectory probably won't do much.

This looks like some sort of IMAP issue, althoug the fact that Mailpile gets stuck saving its config when you ask it to shut down is weird and worrying. What kind of device is this running on? Do you trust the hardware?

**Comment by timorl** [comment]
> What kind of device is this running on? Do you trust the hardware?

It's a Lenovo laptop, not extremely old. I ran smartctl, the drive seems to be fine, output here: https://pastebin.com/4fEQY9UB . I am using btrfs, which supposedly might sometimes be unstable, but so far I had no problems with it. Any other way in which the hardware might be responsible?

I'd still prefer not to redownload everything, so I guess I will keep using the gmail web interface as long as there is some hope. :/

**Comment by BjarniRunar** [comment]
I recently experienced similar problems which were caused by the following cascade of failures: flakey wifi network => hung IMAP connection threads => too many threads! => memory/thread exhaustion at the OS level.

Mailpile was not able to recover from this, I had to kill -9 it.  I've pushed up some fixes to how Mailpile handles timeouts and put a cap on how many threads it will open at once to prevent this. I honestly don't know if this relates to your issue, but I thought I'd mention it.

**Comment by timorl** [comment]
Thank you for the information. I'm afraid it doesn't seem to have changed anything in my case. I think I'll finally just try redownloading everything and hope that helps.

**Comment by BjarniRunar** [comment]
Random thought; have you checked if the machine is low on RAM? Mailpile is quite likely to behave erratically in such cases.

**Comment by timorl** [comment]
I don't think this was an problem with RAM, I turned mailpile on when the only other program running was chromium, so it'd have to use significantly more than 4GB for this to become an issue.

FYI I redownloaded everything into a new profile, so I care significantly less about resolving this bug now. ;P Of course I will still try helping you with figuring it out if you want to, I have a backup of the broken instance as you suggested.
